Mathlib/Algebra/Category/ModuleCat/Presheaf.lean

```diff
@@ -203,4 +234,194 @@ instance : Faithful (toPresheaf R) where
+maps expressed as linear maps to restriction of scalars.
+(See the constructor `PresheafOfModules.mk''`.) -/
+structure BundledCorePresheafOfModules where
+  /-- the datum of a `ModuleCat (R.obj X)` for each object in `Cᵒᵖ` -/
+  obj (X : Cᵒᵖ) : ModuleCat.{v} (R.obj X)
+  /-- the restriction maps as linear maps to restriction of scalars -/
+  map {X Y : Cᵒᵖ} (f : X ⟶ Y) : obj X ⟶ (ModuleCat.restrictScalars (R.map f)).obj (obj Y)
+  /-- `map` is compatible with the identities -/
+  map_id (X : Cᵒᵖ) :
+    map (𝟙 X) = (ModuleCat.restrictScalarsId' (R.map (𝟙 X)) (R.map_id X)).inv.app (obj X)
+  /-- `map` is compatible with the composition -/
+  map_comp {X Y Z : Cᵒᵖ} (f : X ⟶ Y) (g : Y ⟶ Z) :
+    map (f ≫ g) = map f ≫ (ModuleCat.restrictScalars (R.map f)).map (map g) ≫
+      (ModuleCat.restrictScalarsComp' (R.map f) (R.map g) (R.map (f ≫ g))
+        (R.map_comp f g)).inv.app (obj Z)
+
+namespace BundledCorePresheafOfModules
+
+variable (M : BundledCorePresheafOfModules R)
+
+attribute [local simp] map_id map_comp
+
+/-- The obvious map `BundledCorePresheafOfModules R → CorePresheafOfModules R`. -/
+noncomputable def toCorePresheafOfModules : CorePresheafOfModules R where
+  obj X := (M.obj X).carrier
+  map {X Y} f := (ModuleCat.semilinearMapAddEquiv (R.map f) (M.obj X) (M.obj Y)).symm (M.map f)
+
+end BundledCorePresheafOfModules
+
+namespace PresheafOfModules
+
+/-- Constructor for `PresheafOfModules R` based on a collection of objects
+of type `ModuleCat (R.obj X)` for all `X`, and restriction maps expressed
+as linear maps to restriction of scalars, see
+the structure `BundledCorePresheafOfModules`. -/
+noncomputable def mk'' (M : BundledCorePresheafOfModules R) : PresheafOfModules R :=
+  mk' M.toCorePresheafOfModules
+
+@[simp]
+lemma mk''_obj (M : BundledCorePresheafOfModules R) (X : Cᵒᵖ) :
+    (mk'' M).obj X = (M.obj X).carrier := rfl
+
+@[simp]
+lemma mk''_presheaf_map_apply (M : BundledCorePresheafOfModules R)
+    {X Y : Cᵒᵖ} (f : X ⟶ Y) (x : M.obj X) :
+    (mk'' M).presheaf.map f x = M.map f x := rfl
+
+@[simp]
+lemma restrictionApp_mk'' (M : BundledCorePresheafOfModules R) {X Y : Cᵒᵖ} (f : X ⟶ Y) :
+    restrictionApp f (mk'' M) = M.map f := rfl
+
+namespace Hom
+
+variable {P Q : PresheafOfModules R}
+  (app : ∀ X, P.obj X →ₗ[R.obj X] Q.obj X)
+  (naturality : ∀ ⦃X Y : Cᵒᵖ⦄ (f : X ⟶ Y),
+    restrictionApp f P ≫ (ModuleCat.restrictScalars (R.map f)).map (app Y) =
+      ModuleCat.ofHom (app X) ≫ restrictionApp f Q)
+
+/-- A constructor for morphisms in `PresheafOfModules R` that is based on the data
+of a family of linear maps over the various rings `R.obj X`, and for which the
+naturality condition is stated using the restriction of scalars. -/
+def mk'' : P ⟶ Q where
+  hom :=
+    { app := fun X => (app X).toAddMonoidHom
+      naturality := fun X Y f => by
+        ext x
+        exact congr_hom (naturality f) x }
+  map_smul X := (app X).map_smul
+
+@[simp]
+lemma mk''_app : (mk'' app naturality).app = app := rfl
```

Comment:
Maybe this can be grouped together with the `mk'` constructor above?

Reply:
Yes, thanks for the suggestion.